# データサイエンス100本ノック（構造化データ加工編） - Python
# for Google Colab

## 【注意】オリジナル版との変更点
1. Google Colab ではDockerを使えないので、2020.06.18時点の 100knocks-preprocess/docker/work/data にあるCSVファイルをPostgreSQLから入手する代わりに使います。
2. オリジナルのCSVデータ, geocode.csvの'latitude'列名の初めにスペースが入っていたため、それを削除しました。

   オリジナル(100knocks-preprocess ver.1.0): ' latitude' --> 'latitude'
   

4. オリジナルの解答を元に、必要のないlibraryはimportせず、そして必要なlibraryをAzureNotebookでインストールするように最初のセルを改変してあります。
5. また、SQLではなく上記CSVからデータを読み込むように'はじめに'の最初のセルを改変してあります。

## はじめに
- 初めに以下のセルを実行してください
- 必要なライブラリのインポートと~~データベース（PostgreSQL）~~ 100knocks-preprocess/docker/work/data にあるCSVファイルからのデータ読み込みを行います。geocode.csvに変更を加えたため、またgit cloneをするとColabの簡便さを損なうため、noguhiro2002のgithubレポジトリから直接読み込みます。
- pandas等、利用が想定されるライブラリは以下セルでインポートしています
- その他利用したいライブラリがあれば適宜インストールしてください（"!pip install ライブラリ名"でインストールも可能)
- オリジナルの解答例を元に、必要なライブラリーをpipでインストールします。
- 処理は複数回に分けても構いません
- 名前、住所等はダミーデータであり、実在するものではありません

In [17]:
# pipでオリジナルの解答に必要なライブラリーをインストール
#!pip install --upgrade pip
#!pip install -U pandas numpy scikit-learn imbalanced-learn

# pipでオリジナルの解答に必要なライブラリーをインポート
import os
import pandas as pd
import numpy as np
from datetime import datetime, date
from dateutil.relativedelta import relativedelta
import math
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from imblearn.under_sampling import RandomUnderSampler


# データを github/noguhiro2002/100knocks-preprocess/work/data フォルダよりDataframe形式でRead
df_customer = pd.read_csv('customer.csv')
df_category = pd.read_csv('category.csv')
df_product = pd.read_csv('product.csv')
df_receipt = pd.read_csv('receipt.csv')
df_store = pd.read_csv('store.csv')
df_geocode = pd.read_csv('geocode.csv')

C:\Users\Yuka OHATA\Anaconda3\envs\fintalk_100knock\lib\site-packages\IPython\core\interactiveshell.py:2714: DtypeWarning: Columns (4) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


# 演習問題

---
> P-001: レシート明細のデータフレーム（df_receipt）から全項目の先頭10件を表示し、どのようなデータを保有しているか目視で確認せよ。

In [5]:
#headではなく、tailでやると下から
df_receipt.head(10)

sales_ymd  sales_epoch store_cd  receipt_no  receipt_sub_no  \
0   20181103   1541203200   S14006         112               1   
1   20181118   1542499200   S13008        1132               2   
2   20170712   1499817600   S14028        1102               1   
3   20190205   1549324800   S14042        1132               1   
4   20180821   1534809600   S14025        1102               2   
5   20190605   1559692800   S13003        1112               1   
6   20181205   1543968000   S14024        1102               2   
7   20190922   1569110400   S14040        1102               1   
8   20170504   1493856000   S13020        1112               2   
9   20191010   1570665600   S14027        1102               1   

      customer_id  product_cd  quantity  amount  
0  CS006214000001  P070305012         1     158  
1  CS008415000097  P070701017         1      81  
2  CS028414000014  P060101005         1     170  
3  ZZ000000000000  P050301001         1      25  
4  CS025415000050  P060102007         1      90  
5  CS003515000195  P050102002         1     138  
6  CS024514000042  P080101005         1      30  
7  CS040415000178  P070501004         1     128  
8  ZZ000000000000  P071302010         1     770  
9  CS027514000015  P071101003         1     680

---
> P-002: レシート明細のデータフレーム（df_receipt）から売上日（sales_ymd）、顧客ID（customer_id）、商品コード（product_cd）、売上金額（amount）の順に列を指定し、10件表示させよ。

In [ ]:
df_receipt1 =df_receipt[["sales_ymd","customer_id","product_cd","amount"]]
df_receipt1.head(10)

#リストとして入れてしまえば[]は複数でも1つでOK
#例：x=['a','b']
#print(x)

sales_ymd     customer_id  product_cd  amount
0   20181103  CS006214000001  P070305012     158
1   20181118  CS008415000097  P070701017      81
2   20170712  CS028414000014  P060101005     170
3   20190205  ZZ000000000000  P050301001      25
4   20180821  CS025415000050  P060102007      90
5   20190605  CS003515000195  P050102002     138
6   20181205  CS024514000042  P080101005      30
7   20190922  CS040415000178  P070501004     128
8   20170504  ZZ000000000000  P071302010     770
9   20191010  CS027514000015  P071101003     680

---
> P-003: レシート明細のデータフレーム（df_receipt）から売上日（sales_ymd）、顧客ID（customer_id）、商品コード（product_cd）、売上金額（amount）の順に列を指定し、10件表示させよ。ただし、sales_ymdはsales_dateに項目名を変更しながら抽出すること。

In [ ]:
df_receipt2=df_receipt1.rename(columns={'sales_ymd':'sales_data'}).head(10)
df_receipt2

sales_data     customer_id  product_cd  amount
0    20181103  CS006214000001  P070305012     158
1    20181118  CS008415000097  P070701017      81
2    20170712  CS028414000014  P060101005     170
3    20190205  ZZ000000000000  P050301001      25
4    20180821  CS025415000050  P060102007      90
5    20190605  CS003515000195  P050102002     138
6    20181205  CS024514000042  P080101005      30
7    20190922  CS040415000178  P070501004     128
8    20170504  ZZ000000000000  P071302010     770
9    20191010  CS027514000015  P071101003     680

---
> P-004: レシート明細のデータフレーム（df_receipt）から売上日（sales_ymd）、顧客ID（customer_id）、商品コード（product_cd）、売上金額（amount）の順に列を指定し、以下の条件を満たすデータを抽出せよ。
> - 顧客ID（customer_id）が"CS018205000001"

In [ ]:
df_receipt3=df_receipt[['sales_ymd','customer_id','product_cd','amount']].query('customer_id=="CS018205000001"')
df_receipt3

sales_ymd     customer_id  product_cd  amount
36      20180911  CS018205000001  P071401012    2200
9843    20180414  CS018205000001  P060104007     600
21110   20170614  CS018205000001  P050206001     990
27673   20170614  CS018205000001  P060702015     108
27840   20190216  CS018205000001  P071005024     102
28757   20180414  CS018205000001  P071101002     278
39256   20190226  CS018205000001  P070902035     168
58121   20190924  CS018205000001  P060805001     495
68117   20190226  CS018205000001  P071401020    2200
72254   20180911  CS018205000001  P071401005    1100
88508   20190216  CS018205000001  P040101002     218
91525   20190924  CS018205000001  P091503001     280

---
> P-005: レシート明細のデータフレーム（df_receipt）から売上日（sales_ymd）、顧客ID（customer_id）、商品コード（product_cd）、売上金額（amount）の順に列を指定し、以下の条件を満たすデータを抽出せよ。
> - 顧客ID（customer_id）が"CS018205000001"
> - 売上金額（amount）が1,000以上

In [ ]:
#df_receipt4=df_receipt[['sales_ymd','customer_id','product_cd','amount']].query('customer_id=="CS018205000001" & amount>=1000')
#df_receipt4

df_receipt005=df_receipt.copy()
df_receipt005=df_receipt005[['sales_ymd','customer_id','product_cd','amount']]
df_receipt005=[(df_receipt005["customer_id"]=="CS018205000001") & (df_receipt005["amount"]>=1000')]


sales_ymd     customer_id  product_cd  amount
36      20180911  CS018205000001  P071401012    2200
68117   20190226  CS018205000001  P071401020    2200
72254   20180911  CS018205000001  P071401005    1100

---
> P-006: レシート明細データフレーム「df_receipt」から売上日（sales_ymd）、顧客ID（customer_id）、商品コード（product_cd）、売上数量（quantity）、売上金額（amount）の順に列を指定し、以下の条件を満たすデータを抽出せよ。
> - 顧客ID（customer_id）が"CS018205000001"
> - 売上金額（amount）が1,000以上または売上数量（quantity）が5以上

In [ ]:
#df_receipt5=df_receipt[['sales_ymd','customer_id','product_cd','quantity','amount']].query('customer_id=="CS018205000001" & amount>=1000 | quantity>=5')
#df_receipt5

df_receipt006=df_receipt.copy()
df_receipt006=df_receipt006[['sales_ymd','customer_id','product_cd','amount']]
df_receipt006=[(df_receipt006["customer_id"]=="CS018205000001") & (df_receipt006["amount"]>=1000) | (df_receipt006["quantity"]>=5)]

sales_ymd     customer_id  product_cd  quantity  amount
36      20180911  CS018205000001  P071401012         1    2200
9843    20180414  CS018205000001  P060104007         6     600
21110   20170614  CS018205000001  P050206001         5     990
47082   20190610  CS037414000080  P090401002         8     640
68117   20190226  CS018205000001  P071401020         1    2200
72254   20180911  CS018205000001  P071401005         1    1100

---
> P-007: レシート明細のデータフレーム（df_receipt）から売上日（sales_ymd）、顧客ID（customer_id）、商品コード（product_cd）、売上金額（amount）の順に列を指定し、以下の条件を満たすデータを抽出せよ。
> - 顧客ID（customer_id）が"CS018205000001"
> - 売上金額（amount）が1,000以上2,000以下

In [ ]:
#df_receipt6=df_receipt[['sales_ymd','customer_id','product_cd','amount']].query('customer_id=="CS018205000001" & 1000<=amount<=2000')
#df_receipt6

#dfは範囲を繋げて表現できないよ
df_receipt007=df_receipt.copy()
df_receipt007=df_receipt007[['sales_ymd','customer_id','product_cd','amount']]
df_receipt007=[(df_receipt007["customer_id"]=="CS018205000001") & (df_receipt007["amount"]>=1000)& (df_receipt007["amount"]<=2000)]

sales_ymd     customer_id  product_cd  amount
72254   20180911  CS018205000001  P071401005    1100

---
> P-008: レシート明細のデータフレーム（df_receipt）から売上日（sales_ymd）、顧客ID（customer_id）、商品コード（product_cd）、売上金額（amount）の順に列を指定し、以下の条件を満たすデータを抽出せよ。
> - 顧客ID（customer_id）が"CS018205000001"
> - 商品コード（product_cd）が"P071401019"以外

In [ ]:
#df_receipt7=df_receipt[['sales_ymd','customer_id','product_cd','amount']].query('customer_id=="CS018205000001" & product_cd !="P071401019"')
#df_receipt7

df_receipt008=df_receipt.copy()
df_receipt008=df_receipt008[['sales_ymd','customer_id','product_cd','amount']]
df_receipt008=[(df_receipt008["customer_id"]=="CS018205000001")&(df_receipt008["product_cd"]!="P071401019")

sales_ymd     customer_id  product_cd  amount
36      20180911  CS018205000001  P071401012    2200
9843    20180414  CS018205000001  P060104007     600
21110   20170614  CS018205000001  P050206001     990
27673   20170614  CS018205000001  P060702015     108
27840   20190216  CS018205000001  P071005024     102
28757   20180414  CS018205000001  P071101002     278
39256   20190226  CS018205000001  P070902035     168
58121   20190924  CS018205000001  P060805001     495
68117   20190226  CS018205000001  P071401020    2200
72254   20180911  CS018205000001  P071401005    1100
88508   20190216  CS018205000001  P040101002     218
91525   20190924  CS018205000001  P091503001     280

---
> P-009: 以下の処理において、出力結果を変えずにORをANDに書き換えよ。

`df_store.query('not(prefecture_cd == "13" | floor_area > 900)')`

In [ ]:
#df_store.query('prefecture_cd != "13" & floor_area <= 900')

df_store009=df_store.copy()
df_store009=[(df_store009["prefecture_cd"]!=13)&(df_store009["floor_area"]<=900)]

store_cd store_name  prefecture_cd  ... longitude  latitude floor_area
18   S14046       北山田店             14  ...  139.5916  35.56189      831.0
20   S14011      日吉本町店             14  ...  139.6316  35.54655      890.0
38   S12013       習志野店             12  ...  140.0220  35.66122      808.0

[3 rows x 10 columns]

---
> P-010: 店舗データフレーム（df_store）から、店舗コード（store_cd）が"S14"で始まるものだけ全項目抽出し、10件だけ表示せよ。

In [ ]:
#df_store0=df_store.query("store_cd.str.startswith('S14')", engine='python').head(10)
#df_store0

df_store10=df_store[df_store["store_cd"].str.find("S14")>-1].head()
#-1で返すのはfalseのため

store_cd store_name  prefecture_cd  ... longitude  latitude floor_area
2    S14010        菊名店             14  ...  139.6326  35.50049     1732.0
3    S14033       阿久和店             14  ...  139.4961  35.45918     1495.0
4    S14036     相模原中央店             14  ...  139.3716  35.57327     1679.0
7    S14040       長津田店             14  ...  139.4994  35.52398     1548.0
9    S14050      阿久和西店             14  ...  139.4961  35.45918     1830.0
12   S14028       二ツ橋店             14  ...  139.4963  35.46304     1574.0
16   S14012      本牧和田店             14  ...  139.6582  35.42156     1341.0
18   S14046       北山田店             14  ...  139.5916  35.56189      831.0
19   S14022        逗子店             14  ...  139.5789  35.29642     1838.0
20   S14011      日吉本町店             14  ...  139.6316  35.54655      890.0

[10 rows x 10 columns]

---
> P-011: 顧客データフレーム（df_customer）から顧客ID（customer_id）の末尾が1のものだけ全項目抽出し、10件だけ表示せよ。

In [ ]:
df_customer1=df_customer.query("customer_id.str.endswith('1')", engine='python').head(10)
df_customer1

customer_id customer_name  ...  application_date     status_cd
1    CS037613000071         六角 雅彦  ...          20150414  0-00000000-0
3    CS028811000001        堀井 かおり  ...          20160115  0-00000000-0
14   CS040412000191         川井 郁恵  ...          20151101  1-20091025-4
31   CS028314000011        小菅 あおい  ...          20151123  1-20080426-5
56   CS039212000051        藤島 恵梨香  ...          20171121  1-20100215-4
59   CS015412000111         松居 奈月  ...          20150629  0-00000000-0
63   CS004702000041          野島 洋  ...          20170218  0-00000000-0
74   CS041515000001         栗田 千夏  ...          20160422  E-20100803-F
85   CS029313000221        北条 ひかり  ...          20180810  0-00000000-0
102  CS034312000071         望月 奈央  ...          20160106  0-00000000-0

[10 rows x 11 columns]

---
> P-012: 店舗データフレーム（df_store）から横浜市の店舗だけ全項目表示せよ。

In [ ]:
df_store2=df_store.query("address.str.contains('横浜市')", engine='python')
df_store2

#findではなく、containsでも可能
df_store12=df_store[df_store["address"].str.find("横浜市")>-1].head()

store_cd store_name  prefecture_cd  ... longitude  latitude floor_area
2    S14010        菊名店             14  ...  139.6326  35.50049     1732.0
3    S14033       阿久和店             14  ...  139.4961  35.45918     1495.0
7    S14040       長津田店             14  ...  139.4994  35.52398     1548.0
9    S14050      阿久和西店             14  ...  139.4961  35.45918     1830.0
12   S14028       二ツ橋店             14  ...  139.4963  35.46304     1574.0
16   S14012      本牧和田店             14  ...  139.6582  35.42156     1341.0
18   S14046       北山田店             14  ...  139.5916  35.56189      831.0
20   S14011      日吉本町店             14  ...  139.6316  35.54655      890.0
26   S14048      中川中央店             14  ...  139.5758  35.54912     1657.0
40   S14042       新山下店             14  ...  139.6593  35.43894     1044.0
52   S14006       葛が谷店             14  ...  139.5633  35.53573     1886.0

[11 rows x 10 columns]

---
> P-013: 顧客データフレーム（df_customer）から、ステータスコード（status_cd）の先頭がアルファベットのA〜Fで始まるデータを全項目抽出し、10件だけ表示せよ。

In [ ]:
df_customer3=df_customer.query("status_cd.str.contains('^[A-F]',regex=True)", engine='python').head(10)
df_customer3

customer_id customer_name  ...  application_date     status_cd
2   CS031415000172       宇多田 貴美子  ...          20150529  D-20100325-C
6   CS015414000103         奥野 陽子  ...          20150722  B-20100609-B
12  CS011215000048         芦田 沙耶  ...          20150228  C-20100421-9
15  CS029415000023         梅田 里穂  ...          20150610  D-20100918-E
21  CS035415000029         寺沢 真希  ...          20141220  F-20101029-F
32  CS031415000106        宇野 由美子  ...          20150201  F-20100511-E
33  CS029215000025         石倉 美帆  ...          20150708  B-20100820-C
40  CS033605000005         猪股 雄太  ...          20150425  F-20100917-E
44  CS033415000229        板垣 菜々美  ...          20150712  F-20100326-E
53  CS008415000145         黒谷 麻緒  ...          20150829  F-20100622-F

[10 rows x 11 columns]

---
> P-014: 顧客データフレーム（df_customer）から、ステータスコード（status_cd）の末尾が数字の1〜9で終わるデータを全項目抽出し、10件だけ表示せよ。

In [ ]:
df_customer4=df_customer.query("status_cd.str.contains('[1-9]$',regex=True)", engine='python').head(10)
df_customer4.head(10)

customer_id customer_name  ...  application_date     status_cd
4   CS001215000145         田崎 美紀  ...          20170605  6-20090929-2
9   CS033513000180          安斎 遥  ...          20150728  6-20080506-5
12  CS011215000048         芦田 沙耶  ...          20150228  C-20100421-9
14  CS040412000191         川井 郁恵  ...          20151101  1-20091025-4
16  CS009315000023         皆川 文世  ...          20150319  5-20080322-1
22  CS015315000033        福士 璃奈子  ...          20141024  4-20080219-3
23  CS023513000066         神戸 そら  ...          20150915  5-20100524-9
24  CS035513000134         市川 美帆  ...          20150227  8-20100711-9
27  CS001515000263         高松 夏空  ...          20160812  1-20100804-1
28  CS040314000027       鶴田 きみまろ  ...          20150122  2-20080426-4

[10 rows x 11 columns]

---
> P-015: 顧客データフレーム（df_customer）から、ステータスコード（status_cd）の先頭がアルファベットのA〜Fで始まり、末尾が数字の1〜9で終わるデータを全項目抽出し、10件だけ表示せよ。

In [ ]:
#df_customer5=df_customer.query("status_cd.str.contains('^[A-F].+[1-9]$',regex=True)", engine='python').head(10)
df_customer5=df_customer[df_customer["status_cd"].str.contains('^[A-F].+[1-9]$')]
df_customer5

customer_id customer_name  ...  application_date     status_cd
12     CS011215000048         芦田 沙耶  ...          20150228  C-20100421-9
68     CS022513000105        島村 貴美子  ...          20150320  A-20091115-7
71     CS001515000096         水野 陽子  ...          20150614  A-20100724-7
122    CS013615000053         西脇 季衣  ...          20150128  B-20100329-6
144    CS020412000161          小宮 薫  ...          20150822  B-20081021-3
...               ...           ...  ...               ...           ...
21767  CS001412000160          長沢 優  ...          20150907  E-20090412-7
21800  CS005315000066         松井 昌代  ...          20150705  A-20080913-1
21852  CS016513000188         島崎 美幸  ...          20151008  B-20090809-3
21916  CS040615000030         梶原 千夏  ...          20150103  C-20100503-9
21943  CS002515000242        加瀬 はるみ  ...          20170102  A-20100124-7

[620 rows x 11 columns]

---
> P-016: 店舗データフレーム（df_store）から、電話番号（tel_no）が3桁-3桁-4桁のデータを全項目表示せよ。

In [ ]:
df_store6=df_store.query("tel_no.str.contains('[0-9]{3}-[0-9]{3}-[0-9]{4}', regex=True)",engine='python')
df_store6

store_cd store_name  prefecture_cd  ... longitude  latitude floor_area
0    S12014       千草台店             12  ...  140.1180  35.63559     1698.0
1    S13002       国分寺店             13  ...  139.4802  35.70566     1735.0
2    S14010        菊名店             14  ...  139.6326  35.50049     1732.0
3    S14033       阿久和店             14  ...  139.4961  35.45918     1495.0
4    S14036     相模原中央店             14  ...  139.3716  35.57327     1679.0
7    S14040       長津田店             14  ...  139.4994  35.52398     1548.0
9    S14050      阿久和西店             14  ...  139.4961  35.45918     1830.0
11   S13052        森野店             13  ...  139.4383  35.55293     1087.0
12   S14028       二ツ橋店             14  ...  139.4963  35.46304     1574.0
16   S14012      本牧和田店             14  ...  139.6582  35.42156     1341.0
18   S14046       北山田店             14  ...  139.5916  35.56189      831.0
19   S14022        逗子店             14  ...  139.5789  35.29642     1838.0
20   S14011      日吉本町店             14  ...  139.6316  35.54655      890.0
21   S13016       小金井店             13  ...  139.5094  35.70018     1399.0
22   S14034      川崎野川店             14  ...  139.5998  35.57693     1318.0
26   S14048      中川中央店             14  ...  139.5758  35.54912     1657.0
27   S12007        佐倉店             12  ...  140.1452  35.71872     1895.0
28   S14026     辻堂西海岸店             14  ...  139.4466  35.32464     1732.0
29   S13041       八王子店             13  ...  139.4235  35.63787      810.0
31   S14049      川崎大師店             14  ...  139.7327  35.53759      962.0
32   S14023        川崎店             14  ...  139.7028  35.53599     1804.0
33   S13018        清瀬店             13  ...  139.5178  35.76885     1220.0
35   S14027       南藤沢店             14  ...  139.4896  35.33762     1521.0
36   S14021       伊勢原店             14  ...  139.3129  35.40169      962.0
37   S14047       相模原店             14  ...  139.3748  35.55959     1047.0
38   S12013       習志野店             12  ...  140.0220  35.66122      808.0
40   S14042       新山下店             14  ...  139.6593  35.43894     1044.0
42   S12030        八幡店             12  ...  139.9240  35.72318     1162.0
44   S14025        大和店             14  ...  139.4680  35.43414     1011.0
45   S14045        厚木店             14  ...  139.3651  35.44182      980.0
47   S12029        東野店             12  ...  139.8968  35.65086     1101.0
49   S12053        高洲店             12  ...  139.9176  35.63755     1555.0
51   S14024        三田店             14  ...  139.5424  35.60770      972.0
52   S14006       葛が谷店             14  ...  139.5633  35.53573     1886.0

[34 rows x 10 columns]

---
> P-17: 顧客データフレーム（df_customer）を生年月日（birth_day）で高齢順にソートし、先頭10件を全項目表示せよ。

In [ ]:
df_customer17=df_customer.sort_values('birth_day').head(10)
df_customer17

customer_id customer_name  ...  application_date     status_cd
18817  CS003813000014        村山 菜々美  ...          20160214  0-00000000-0
12328  CS026813000004         吉村 朝陽  ...          20150723  0-00000000-0
15682  CS018811000003         熊沢 美里  ...          20150403  0-00000000-0
15302  CS027803000004         内村 拓郎  ...          20151227  0-00000000-0
1681   CS013801000003         天野 拓郎  ...          20160120  0-00000000-0
7511   CS001814000022         鶴田 里穂  ...          20161012  A-20090415-7
2378   CS016815000002         山元 美紀  ...          20150629  C-20090923-C
4680   CS009815000003         中田 里穂  ...          20150421  D-20091021-E
16070  CS005813000015        金谷 恵梨香  ...          20150506  0-00000000-0
6305   CS012813000013         宇野 南朋  ...          20150712  0-00000000-0

[10 rows x 11 columns]

---
> P-18: 顧客データフレーム（df_customer）を生年月日（birth_day）で若い順にソートし、先頭10件を全項目表示せよ。

In [ ]:
df_customer18=df_customer.sort_values('birth_day',ascending=False).head(10)
df_customer18

customer_id customer_name  ...  application_date     status_cd
15639  CS035114000004         大村 美里  ...          20150619  6-20091205-6
7468   CS022103000002        福山 はじめ  ...          20160909  0-00000000-0
10745  CS002113000009        柴田 真悠子  ...          20160304  0-00000000-0
19811  CS004115000014         松井 京子  ...          20161120  1-20081231-1
7039   CS002114000010          山内 遥  ...          20160920  6-20100510-1
3670   CS025115000002         小柳 夏希  ...          20160116  D-20100913-D
12493  CS002113000025        広末 まなみ  ...          20171030  0-00000000-0
15977  CS033112000003         長野 美紀  ...          20150606  0-00000000-0
5716   CS007115000006          福岡 瞬  ...          20151118  F-20101016-F
15097  CS014113000008         矢口 莉緒  ...          20150622  3-20091108-6

[10 rows x 11 columns]

---
> P-19: レシート明細データフレーム（df_receipt）に対し、1件あたりの売上金額（amount）が高い順にランクを付与し、先頭10件を抽出せよ。項目は顧客ID（customer_id）、売上金額（amount）、付与したランクを表示させること。なお、売上金額（amount）が等しい場合は同一順位を付与するものとする。

In [ ]:
df_receipt19 = pd.concat([df_receipt[['customer_id', 'amount']] 
                     ,df_receipt['amount'].rank(method='min', ascending=False)], axis=1)
df_receipt19.columns = ['customer_id', 'amount', 'ranking']
df_receipt19.sort_values('ranking', ascending=True).head(10)

customer_id  amount  ranking
1202   CS011415000006   10925      1.0
62317  ZZ000000000000    6800      2.0
54095  CS028605000002    5780      3.0
4632   CS015515000034    5480      4.0
72747  ZZ000000000000    5480      4.0
10320  ZZ000000000000    5480      4.0
97294  CS021515000089    5440      7.0
28304  ZZ000000000000    5440      7.0
92246  CS009415000038    5280      9.0
68553  CS040415000200    5280      9.0

---
> P-020: レシート明細データフレーム（df_receipt）に対し、1件あたりの売上金額（amount）が高い順にランクを付与し、先頭10件を抽出せよ。項目は顧客ID（customer_id）、売上金額（amount）、付与したランクを表示させること。なお、売上金額（amount）が等しい場合でも別順位を付与すること。

In [ ]:
df_receipt19 = pd.concat([df_receipt[['customer_id', 'amount']] 
                     ,df_receipt['amount'].rank(method='first', ascending=False)], axis=1)
df_receipt19.columns = ['customer_id', 'amount', 'ranking']
df_receipt19.sort_values('ranking', ascending=True).head(10)

customer_id  amount  ranking
1202   CS011415000006   10925      1.0
62317  ZZ000000000000    6800      2.0
54095  CS028605000002    5780      3.0
4632   CS015515000034    5480      4.0
10320  ZZ000000000000    5480      5.0
72747  ZZ000000000000    5480      6.0
28304  ZZ000000000000    5440      7.0
97294  CS021515000089    5440      8.0
596    CS015515000083    5280      9.0
11275  CS017414000114    5280     10.0

---
> P-021: レシート明細データフレーム（df_receipt）に対し、件数をカウントせよ。

In [ ]:
df_receipt.count()

sales_ymd         104681
sales_epoch       104681
store_cd          104681
receipt_no        104681
receipt_sub_no    104681
customer_id       104681
product_cd        104681
quantity          104681
amount            104681
dtype: int64

---
> P-022: レシート明細データフレーム（df_receipt）の顧客ID（customer_id）に対し、ユニーク件数をカウントせよ。

In [ ]:
len(df_receipt['customer_id'].unique())

8307

---
> P-023: レシート明細データフレーム（df_receipt）に対し、店舗コード（store_cd）ごとに売上金額（amount）と売上数量（quantity）を合計せよ。

In [ ]:
df_receipt.groupby('store_cd')[['amount','quantity']].sum()

amount  quantity
store_cd                  
S12007    638761      2099
S12013    787513      2425
S12014    725167      2358
S12029    794741      2555
S12030    684402      2403
S13001    811936      2347
S13002    727821      2340
S13003    764294      2197
S13004    779373      2390
S13005    629876      2004
S13008    809288      2491
S13009    808870      2486
S13015    780873      2248
S13016    793773      2432
S13017    748221      2376
S13018    790535      2562
S13019    827833      2541
S13020    796383      2383
S13031    705968      2336
S13032    790501      2491
S13035    715869      2219
S13037    693087      2344
S13038    708884      2337
S13039    611888      1981
S13041    728266      2233
S13043    587895      1881
S13044    520764      1729
S13051    107452       354
S13052    100314       250
S14006    712839      2284
S14010    790361      2290
S14011    805724      2434
S14012    720600      2412
S14021    699511      2231
S14022    651328      2047
S14023    727630      2258
S14024    736323      2417
S14025    755581      2394
S14026    824537      2503
S14027    714550      2303
S14028    786145      2458
S14033    725318      2282
S14034    653681      2024
S14036    203694       635
S14040    701858      2233
S14042    534689      1935
S14045    458484      1398
S14046    412646      1354
S14047    338329      1041
S14048    234276       769
S14049    230808       788
S14050    167090       580

---
> P-024: レシート明細データフレーム（df_receipt）に対し、顧客ID（customer_id）ごとに最も新しい売上日（sales_ymd）を求め、10件表示せよ。

In [ ]:
df24=df_receipt.copy()
df24["sales_ymd"]=pd.to_datetime(df24["sales_ymd"],format="%Y%m%d")
df24.groupby('customer_id')['sales_ymd'].max().reset_index().head(10)

customer_id  sales_ymd
0  CS001113000004 2019-03-08
1  CS001114000005 2019-07-31
2  CS001115000010 2019-04-05
3  CS001205000004 2019-06-25
4  CS001205000006 2019-02-24
5  CS001211000025 2019-03-22
6  CS001212000027 2017-01-27
7  CS001212000031 2018-09-06
8  CS001212000046 2017-08-11
9  CS001212000070 2019-10-18

---
> P-025: レシート明細データフレーム（df_receipt）に対し、顧客ID（customer_id）ごとに最も古い売上日（sales_ymd）を求め、10件表示せよ。

In [ ]:
df25=df_receipt.copy()
df25["sales_ymd"]=pd.to_datetime(df25["sales_ymd"],format="%Y%m%d")
df25.groupby('customer_id').sales_ymd.min().reset_index().head(10)

customer_id  sales_ymd
0  CS001113000004 2019-03-08
1  CS001114000005 2018-05-03
2  CS001115000010 2017-12-28
3  CS001205000004 2017-09-14
4  CS001205000006 2018-02-07
5  CS001211000025 2019-03-22
6  CS001212000027 2017-01-27
7  CS001212000031 2018-09-06
8  CS001212000046 2017-08-11
9  CS001212000070 2019-10-18

---
> P-026: レシート明細データフレーム（df_receipt）に対し、顧客ID（customer_id）ごとに最も新しい売上日（sales_ymd）と古い売上日を求め、両者が異なるデータを10件表示せよ。

In [ ]:
df26_max=df_receipt.groupby('customer_id').sales_ymd.max()
df26_min=df_receipt.groupby('customer_id').sales_ymd.min()
df26=pd.merge(df26_max, df26_min, on='customer_id')
df26=df26.rename(columns={'sales_ymd_x': 'sales_ymd_max', 'sales_ymd_y': 'sales_ymd_min'})
#df26.query('sales_ymd_max != sales_ymd_min').head(10)
df26[df26["sales_ymd_max"]!=df26["sales_ymd_min"]]

sales_ymd_max  sales_ymd_min
customer_id                                 
CS001114000005       20190731       20180503
CS001115000010       20190405       20171228
CS001205000004       20190625       20170914
CS001205000006       20190224       20180207
CS001214000009       20190902       20170306
...                       ...            ...
CS049514000003       20190731       20190207
CS049605000005       20190917       20190423
CS049613000002       20181228       20181207
CS050415000007       20191013       20190720
ZZ000000000000       20191031       20170101

[5546 rows x 2 columns]

In [ ]:
df26_max=df_receipt.groupby('customer_id').sales_ymd.max()
df26_min=df_receipt.groupby('customer_id').sales_ymd.min()
df26_max.name="sales_ymd_max"
df26_min.name="sales_ymd_min"
#こんかち(同じカラム名があるとき勝手に合体してくれる)
#axis=1で横並び。そうでないと縦になる
#[]は塊をリストに渡しているから、複数でも[[]]にしなくてOK
pd.concat([df26_max,df26_min],axis=1)

sales_ymd_max  sales_ymd_min
customer_id                                 
CS001113000004       20190308       20190308
CS001114000005       20190731       20180503
CS001115000010       20190405       20171228
CS001205000004       20190625       20170914
CS001205000006       20190224       20180207
...                       ...            ...
CS051513000004       20190719       20190719
CS051515000002       20191025       20191025
CS052212000002       20191017       20191017
CS052514000001       20190822       20190822
ZZ000000000000       20191031       20170101

[8307 rows x 2 columns]

---
> P-027: レシート明細データフレーム（df_receipt）に対し、店舗コード（store_cd）ごとに売上金額（amount）の平均を計算し、降順でTOP5を表示せよ。

In [ ]:
df27=df_receipt.groupby('store_cd').amount.mean().reset_index()
df27.sort_values('amount',ascending=False).head(5)

store_cd      amount
28   S13052  402.867470
12   S13015  351.111960
7    S13003  350.915519
30   S14010  348.791262
5    S13001  348.470386

---
> P-028: レシート明細データフレーム（df_receipt）に対し、店舗コード（store_cd）ごとに売上金額（amount）の中央値を計算し、降順でTOP5を表示せよ。

In [ ]:
df28=df_receipt.groupby('store_cd').amount.median().reset_index()
df28.sort_values('amount',ascending=False).head(5)

store_cd  amount
28   S13052     190
30   S14010     188
51   S14050     185
44   S14040     180
7    S13003     180

---
> P-029: レシート明細データフレーム（df_receipt）に対し、店舗コード（store_cd）ごとに商品コード（product_cd）の最頻値を求めよ。

In [ ]:
df_receipt.groupby('store_cd').product_cd.apply(lambda x: x.mode()).reset_index()
list(df_receipt.groupby('store_cd').product_cd)[0][1]
#groupbyが提供できるメソッドは決まっている→applyでmode関数を適用する

31        P080104025
211       P071401011
222       P060503001
235       P070308044
262       P060102005
             ...    
104191    P071004020
104241    P040102005
104429    P071401009
104456    P071001060
104612    P060303001
Name: product_cd, Length: 2076, dtype: object

---
> P-030: レシート明細データフレーム（df_receipt）に対し、店舗コード（store_cd）ごとに売上金額（amount）の標本分散を計算し、降順でTOP5を表示せよ。

In [ ]:
df30=df_receipt.groupby('store_cd').amount.var(ddof=0).reset_index()
df30.sort_values('amount',ascending=False).head(5)

store_cd         amount
28   S13052  440088.701311
31   S14011  306314.558164
42   S14034  296920.081011
5    S13001  295431.993329
12   S13015  295294.361116

---
> P-031: レシート明細データフレーム（df_receipt）に対し、店舗コード（store_cd）ごとに売上金額（amount）の標本標準偏差を計算し、降順でTOP5を表示せよ。

In [ ]:
df31=df_receipt.groupby('store_cd').amount.std(ddof=0).reset_index()
df31.sort_values('amount',ascending=False).head(5)

store_cd      amount
28   S13052  663.391816
31   S14011  553.456916
42   S14034  544.903736
5    S13001  543.536561
12   S13015  543.409938

---
> P-032: レシート明細データフレーム（df_receipt）の売上金額（amount）について、25％刻みでパーセンタイル値を求めよ。

In [6]:
df_receipt["amount"].quantile([0.0,0.25,0.5,0.75,1.0])

0.00       10.0
0.25      102.0
0.50      170.0
0.75      288.0
1.00    10925.0
Name: amount, dtype: float64

---
> P-033: レシート明細データフレーム（df_receipt）に対し、店舗コード（store_cd）ごとに売上金額（amount）の平均を計算し、330以上のものを抽出せよ。

In [10]:
#df_receipt.groupby('store_cd').amount.mean().reset_index().query('amount>=330')

df_receipt33 = df_receipt.groupby("store_cd")[["amount"]].mean()
df_receipt33[df_receipt33["amount"] >= 330]

amount
store_cd            
S12013    330.194130
S13001    348.470386
S13003    350.915519
S13004    330.943949
S13015    351.111960
S13019    330.208616
S13020    337.879932
S13052    402.867470
S14010    348.791262
S14011    335.718333
S14026    332.340588
S14045    330.082073
S14047    330.077073

---
> P-034: レシート明細データフレーム（df_receipt）に対し、顧客ID（customer_id）ごとに売上金額（amount）を合計して全顧客の平均を求めよ。ただし、顧客IDが"Z"から始まるのものは非会員を表すため、除外して計算すること。


In [ ]:
df_receipt.query('not customer_id.str.startswith("Z")', engine='python').groupby('customer_id').amount.mean().reset_index()

customer_id      amount
0     CS001113000004  649.000000
1     CS001114000005  156.500000
2     CS001115000010  507.333333
3     CS001205000004  198.800000
4     CS001205000006  278.083333
...              ...         ...
8301  CS051212000001  168.000000
8302  CS051513000004  275.500000
8303  CS051515000002  132.500000
8304  CS052212000002   96.000000
8305  CS052514000001   89.000000

[8306 rows x 2 columns]

---
> P-035: レシート明細データフレーム（df_receipt）に対し、顧客ID（customer_id）ごとに売上金額（amount）を合計して全顧客の平均を求め、平均以上に買い物をしている顧客を抽出せよ。ただし、顧客IDが"Z"から始まるのものは非会員を表すため、除外して計算すること。なお、データは10件だけ表示させれば良い。

In [ ]:
df35=df_receipt.query('not customer_id.str.startswith("Z")', engine='python').groupby('customer_id').amount.mean().reset_index()
df35.query('amount>=amount.mean()').head(10)

customer_id       amount
0   CS001113000004   649.000000
2   CS001115000010   507.333333
13  CS001214000009   334.642857
14  CS001214000017   344.333333
17  CS001214000052   469.916667
21  CS001215000040   349.600000
24  CS001215000097  1144.000000
26  CS001215000145   437.500000
27  CS001215000158   337.250000
29  CS001304000004   624.000000

---
> P-036: レシート明細データフレーム（df_receipt）と店舗データフレーム（df_store）を内部結合し、レシート明細データフレームの全項目と店舗データフレームの店舗名（store_name）を10件表示させよ。

In [ ]:
df36=df_receipt.merge(df_store)
df36_1=df36[["sales_ymd","sales_epoch","store_cd","receipt_no","receipt_sub_no","customer_id","product_cd","quantity","amount","store_name"]]
df36_1.head(10)

sales_ymd  sales_epoch store_cd  ...  quantity  amount store_name
0   20181103   1541203200   S14006  ...         1     158       葛が谷店
1   20181116   1542326400   S14006  ...         1      48       葛が谷店
2   20170118   1484697600   S14006  ...         1     220       葛が谷店
3   20190524   1558656000   S14006  ...         1      80       葛が谷店
4   20190419   1555632000   S14006  ...         1     148       葛が谷店
5   20181119   1542585600   S14006  ...         1      88       葛が谷店
6   20171211   1512950400   S14006  ...         1      80       葛が谷店
7   20191021   1571616000   S14006  ...         1     405       葛が谷店
8   20170710   1499644800   S14006  ...         1     330       葛が谷店
9   20190805   1564963200   S14006  ...         1     115       葛が谷店

[10 rows x 10 columns]

---
> P-037: 商品データフレーム（df_product）とカテゴリデータフレーム（df_category）を内部結合し、商品データフレームの全項目とカテゴリデータフレームの小区分名（category_small_name）を10件表示させよ。

In [ ]:
df37=df_product.merge(df_category)
df37_1=df37[['product_cd','category_major_cd','category_medium_cd','category_small_cd','unit_price','unit_cost','category_small_name']]
df37_1.head(10)

product_cd  category_major_cd  ...  unit_cost  category_small_name
0  P040101001                  4  ...      149.0                  弁当類
1  P040101002                  4  ...      164.0                  弁当類
2  P040101003                  4  ...      173.0                  弁当類
3  P040101004                  4  ...      186.0                  弁当類
4  P040101005                  4  ...      201.0                  弁当類
5  P040101006                  4  ...      224.0                  弁当類
6  P040101007                  4  ...      254.0                  弁当類
7  P040101008                  4  ...      315.0                  弁当類
8  P040101009                  4  ...      374.0                  弁当類
9  P040101010                  4  ...      435.0                  弁当類

[10 rows x 7 columns]

---
> P-038: 顧客データフレーム（df_customer）とレシート明細データフレーム（df_receipt）から、各顧客ごとの売上金額合計を求めよ。ただし、買い物の実績がない顧客については売上金額を0として表示させること。また、顧客は性別コード（gender_cd）が女性（1）であるものを対象とし、非会員（顧客IDが'Z'から始まるもの）は除外すること。なお、結果は10件だけ表示させれば良い。

In [27]:
df38_1[df38_1['customer_id'].str.startswith("Z")]

customer_id  amount  gender_cd
48     ZZ000000000000      25        NaN
49     ZZ000000000000     770        NaN
50     ZZ000000000000     115        NaN
51     ZZ000000000000     100        NaN
52     ZZ000000000000     268        NaN
...               ...     ...        ...
39042  ZZ000000000000     308        NaN
39043  ZZ000000000000     100        NaN
39044  ZZ000000000000      40        NaN
39045  ZZ000000000000     218        NaN
39046  ZZ000000000000     138        NaN

[38999 rows x 3 columns]

In [36]:
df38=df_customer.merge(df_receipt,how='right',left_on='customer_id',right_on='customer_id').fillna({"amount":0})
df38_1=df38[['customer_id','amount','gender_cd']]
df38_2=df38_1.groupby(['customer_id','gender_cd']).amount.sum().reset_index()
#df38_3=df38_2.query('gender_cd == "1" and not customer_id.str.startswith("Z")', engine='python').fillna(0)
df38_2[(df38_2['gender_cd']==1) & (~df38_2['customer_id'].str.startswith("Z"))].groupby(['customer_id','gender_cd'])[["amount"]].sum()
#df38_2[~df38_2['customer_id'].str.startswith("Z")]

amount
customer_id    gender_cd        
CS001113000004 1.0          1298
CS001114000005 1.0           626
CS001115000010 1.0          3044
CS001211000025 1.0           456
CS001212000027 1.0           448
...                          ...
CS051212000001 1.0           336
CS051513000004 1.0           551
CS051515000002 1.0           265
CS052212000002 1.0           192
CS052514000001 1.0           178

[7618 rows x 1 columns]

---
> P-039: レシート明細データフレーム（df_receipt）から売上日数の多い顧客の上位20件と、売上金額合計の多い顧客の上位20件を抽出し、完全外部結合せよ。ただし、非会員（顧客IDが'Z'から始まるもの）は除外すること。

In [ ]:
#完全外部結合？？？？？？
#レシート
df_r1=df_receipt.groupby('customer_id').amount.sum().reset_index()
df_r2=df_r1.query('not customer_id.str.startswith("Z")', engine='python')
df_r3=df_r2.sort_values('amount',ascending=False).head(20)

#合計金額
df_s1=df_receipt.groupby(['customer_id','sales_ymd']).amount.sum().reset_index()
df_s2=df_s1.query('not customer_id.str.startswith("Z")', engine='python')
df_s3=df_s2.sort_values('amount',ascending=False).head(20)

df39=df_r3.merge(df_s3,on='customer_id',how='outer')
df39

customer_id  amount_x   sales_ymd  amount_y
0   CS017415000097   23086.0         NaN       NaN
1   CS015415000185   20153.0         NaN       NaN
2   CS031414000051   19202.0         NaN       NaN
3   CS028415000007   19127.0         NaN       NaN
4   CS001605000009   18925.0  20190525.0    5860.0
5   CS010214000010   18585.0         NaN       NaN
6   CS016415000141   18372.0         NaN       NaN
7   CS006515000023   18372.0         NaN       NaN
8   CS011414000106   18338.0  20191001.0    6380.0
9   CS038415000104   17847.0         NaN       NaN
10  CS035414000024   17615.0         NaN       NaN
11  CS021515000089   17580.0  20190209.0    5753.0
12  CS032414000072   16563.0  20170224.0    6380.0
13  CS016415000101   16348.0         NaN       NaN
14  CS011415000006   16094.0  20190903.0   11085.0
15  CS034415000047   16083.0         NaN       NaN
16  CS007514000094   15735.0         NaN       NaN
17  CS009414000059   15492.0         NaN       NaN
18  CS030415000034   15468.0  20170326.0    5860.0
19  CS015515000034   15300.0  20170721.0    5580.0
20  CS018415000125       NaN  20170216.0    7710.0
21  CS020515000102       NaN  20180511.0    7680.0
22  CS035414000008       NaN  20170101.0    7473.0
23  CS005514000202       NaN  20181108.0    7200.0
24  CS038415000071       NaN  20180714.0    6410.0
25  CS028515000172       NaN  20181216.0    6380.0
26  CS040515000225       NaN  20171027.0    6360.0
27  CS039515000090       NaN  20191010.0    6260.0
28  CS040415000035       NaN  20170314.0    6200.0
29  CS034515000173       NaN  20171130.0    6080.0
30  CS030515000189       NaN  20181019.0    6048.0
31  CS028605000002       NaN  20170921.0    5880.0
32  CS035414000047       NaN  20170710.0    5786.0

---
> P-040: 全ての店舗と全ての商品を組み合わせると何件のデータとなるか調査したい。店舗（df_store）と商品（df_product）を直積した件数を計算せよ。

In [ ]:
len(df_store)*len(df_product)

531590

---
> P-041: レシート明細データフレーム（df_receipt）の売上金額（amount）を日付（sales_ymd）ごとに集計し、前日からの売上金額増減を計算せよ。なお、計算結果は10件表示すればよい。

In [9]:
#df41a=df_receipt.groupby('sales_ymd').amount.sum().reset_index()
#df41b = df_receipt.groupby('sales_ymd').amount.sum().reset_index()
#df41b_1= df41b.shift(-1).reset_index()

#1つずらして作ればいいのかな？というイメージは出ましたが、わかりませんでした
#concat？？
df_sales_amount_by_date = df_receipt[['sales_ymd', 'amount']].groupby('sales_ymd').sum().reset_index()
df_sales_amount_by_date = pd.concat([df_sales_amount_by_date, df_sales_amount_by_date.shift()], axis=1)
df_sales_amount_by_date.columns = ['sales_ymd','amount','lag_ymd','lag_amount']
df_sales_amount_by_date['diff_amount'] = df_sales_amount_by_date['amount'] - df_sales_amount_by_date['lag_amount']
df_sales_amount_by_date.head(10)

sales_ymd  amount     lag_ymd  lag_amount  diff_amount
0   20170101   33723         NaN         NaN          NaN
1   20170102   24165  20170101.0     33723.0      -9558.0
2   20170103   27503  20170102.0     24165.0       3338.0
3   20170104   36165  20170103.0     27503.0       8662.0
4   20170105   37830  20170104.0     36165.0       1665.0
5   20170106   32387  20170105.0     37830.0      -5443.0
6   20170107   23415  20170106.0     32387.0      -8972.0
7   20170108   24737  20170107.0     23415.0       1322.0
8   20170109   26718  20170108.0     24737.0       1981.0
9   20170110   20143  20170109.0     26718.0      -6575.0

---
> P-042: レシート明細データフレーム（df_receipt）の売上金額（amount）を日付（sales_ymd）ごとに集計し、各日付のデータに対し、１日前、２日前、３日前のデータを結合せよ。結果は10件表示すればよい。

In [90]:
df_sales_amount_by_date = df_receipt[['sales_ymd', 'amount']].groupby('sales_ymd').sum().reset_index()
df_sales_amount_by_date = pd.concat([df_sales_amount_by_date.shift()], axis=1)
df_sales_amount_by_date_1 = pd.concat([df_sales_amount_by_date.shift(-1),df_sales_amount_by_date], axis=1)

col = df_sales_amount_by_date_1.columns.values
col[1]='amount'
col[2]='no'
col[3]='1bf_amount'

In [118]:
df_sales_amount_by_date_2 = pd.concat([df_sales_amount_by_date.shift(-2),df_sales_amount_by_date], axis=1)

col = df_sales_amount_by_date_2.columns.values
col[1]='1bf_amount'
col[2]='no'
col[3]='2bf_amount'

df_sales_amount_by_date_2=df_sales_amount_by_date_2[['sales_ymd','1bf_amount','2bf_amount']]

In [121]:
df_sales_amount_by_date_3 = pd.concat([df_sales_amount_by_date_1.shift(-2),df_sales_amount_by_date_1], axis=1)

col = df_sales_amount_by_date_3.columns.values
col[1]='3bf_amount'
col[2]='no'
col[3]='2bf_amount'
col[4]='no'
col[5]='1bf_amount'
col[6]='no'
col[7]='no'

df_sales_amount_by_date_3=df_sales_amount_by_date_3[['sales_ymd','1bf_amount','2bf_amount','3bf_amount']]
df_sales_amount_by_date_3.head(10)

sales_ymd  1bf_amount  2bf_amount  3bf_amount
0  20170103.0     33723.0     24165.0     27503.0
1  20170104.0     24165.0     27503.0     36165.0
2  20170105.0     27503.0     36165.0     37830.0
3  20170106.0     36165.0     37830.0     32387.0
4  20170107.0     37830.0     32387.0     23415.0
5  20170108.0     32387.0     23415.0     24737.0
6  20170109.0     23415.0     24737.0     26718.0
7  20170110.0     24737.0     26718.0     20143.0
8  20170111.0     26718.0     20143.0     24287.0
9  20170112.0     20143.0     24287.0     23526.0

---
> P-043： レシート明細データフレーム（df_receipt）と顧客データフレーム（df_customer）を結合し、性別（gender）と年代（ageから計算）ごとに売上金額（amount）を合計した売上サマリデータフレーム（df_sales_summary）を作成せよ。性別は0が男性、1が女性、9が不明を表すものとする。
>
> ただし、項目構成は年代、女性の売上金額、男性の売上金額、性別不明の売上金額の4項目とすること（縦に年代、横に性別のクロス集計）。また、年代は10歳ごとの階級とすること。

In [127]:
#カンニング
#そもそもたてもちとは？？
df_tmp = pd.merge(df_receipt, df_customer, how ='inner', on="customer_id")
df_tmp['era'] = df_tmp['age'].apply(lambda x: math.floor(x / 10) * 10)
df_sales_summary = pd.pivot_table(df_tmp, index='era', columns='gender_cd', values='amount', aggfunc='sum').reset_index()
df_sales_summary.columns = ['era', 'male', 'female', 'unknown']
df_sales_summary

era      male     female   unknown
0   10    1591.0   149836.0    4317.0
1   20   72940.0  1363724.0   44328.0
2   30  177322.0   693047.0   50441.0
3   40   19355.0  9320791.0  483512.0
4   50   54320.0  6685192.0  342923.0
5   60  272469.0   987741.0   71418.0
6   70   13435.0    29764.0    2427.0
7   80   46360.0   262923.0    5111.0
8   90       NaN     6260.0       NaN

---
> P-044： 前設問で作成した売上サマリデータフレーム（df_sales_summary）は性別の売上を横持ちさせたものであった。このデータフレームから性別を縦持ちさせ、年代、性別コード、売上金額の3項目に変換せよ。ただし、性別コードは男性を'00'、女性を'01'、不明を'99'とする。

In [128]:
df_sales_summary = df_sales_summary.set_index('era'). \
        stack().reset_index().replace({'female':'01',
                                        'male':'00',
                                        'unknown':'99'}).rename(columns={'level_1':'gender_cd', 0: 'amount'})


---
> P-045: 顧客データフレーム（df_customer）の生年月日（birth_day）は日付型（Date）でデータを保有している。これをYYYYMMDD形式の文字列に変換し、顧客ID（customer_id）とともに抽出せよ。データは10件を抽出すれば良い。

In [ ]:
df45=df_customer.copy()

df45['birth_day']=pd.to_datetime(df45['birth_day']).dt.strftime('%Y%m%d')
df45[['customer_id','birth_day']].head(10)

customer_id birth_day
0  CS021313000114  19810429
1  CS037613000071  19520401
2  CS031415000172  19761004
3  CS028811000001  19330327
4  CS001215000145  19950329
5  CS020401000016  19740915
6  CS015414000103  19770809
7  CS029403000008  19730817
8  CS015804000004  19310502
9  CS033513000180  19620711

---
> P-046: 顧客データフレーム（df_customer）の申し込み日（application_date）はYYYYMMD形式の文字列型でデータを保有している。これを日付型（dateやdatetime）に変換し、顧客ID（customer_id）とともに抽出せよ。データは10件を抽出すれば良い。

In [ ]:
df46=df_customer.copy()
df46['application_date']=pd.to_datetime(df46['application_date'],format=('%Y%m%d'))
df46[['customer_id','application_date']].head(10)

#df46['application_date']

0       2015-09-05
1       2015-04-14
2       2015-05-29
3       2016-01-15
4       2017-06-05
           ...    
21966   2017-11-10
21967   2015-03-13
21968   2015-04-06
21969   2016-02-06
21970   2015-04-24
Name: application_date, Length: 21971, dtype: datetime64[ns]

---
> P-047: レシート明細データフレーム（df_receipt）の売上日（sales_ymd）はYYYYMMDD形式の数値型でデータを保有している。これを日付型（dateやdatetime）に変換し、レシート番号(receipt_no)、レシートサブ番号（receipt_sub_no）とともに抽出せよ。データは10件を抽出すれば良い。

In [ ]:
df47=df_receipt.copy()
df47['sales_ymd']=pd.to_datetime(df47['sales_ymd'],format=('%Y%m%d'))
df47[['receipt_no','receipt_sub_no','sales_ymd']].head(10)

#df47['sales_ymd']

0        2018-11-03
1        2018-11-18
2        2017-07-12
3        2019-02-05
4        2018-08-21
            ...    
104676   2018-02-21
104677   2019-09-11
104678   2017-03-11
104679   2017-03-31
104680   2019-04-23
Name: sales_ymd, Length: 104681, dtype: datetime64[ns]

---
> P-048: レシート明細データフレーム（df_receipt）の売上エポック秒（sales_epoch）は数値型のUNIX秒でデータを保有している。これを日付型（dateやdatetime）に変換し、レシート番号(receipt_no)、レシートサブ番号（receipt_sub_no）とともに抽出せよ。データは10件を抽出すれば良い。

In [ ]:
df48=df_receipt.copy()
df48['sales_epoch']=pd.to_datetime(df48['sales_epoch'])
df48[['receipt_no','receipt_sub_no','sales_epoch']].head(10)

#df48['sales_epoch']

0        1970-01-01 00:00:01.541203200
1        1970-01-01 00:00:01.542499200
2        1970-01-01 00:00:01.499817600
3        1970-01-01 00:00:01.549324800
4        1970-01-01 00:00:01.534809600
                      ...             
104676   1970-01-01 00:00:01.519171200
104677   1970-01-01 00:00:01.568160000
104678   1970-01-01 00:00:01.489190400
104679   1970-01-01 00:00:01.490918400
104680   1970-01-01 00:00:01.555977600
Name: sales_epoch, Length: 104681, dtype: datetime64[ns]

---
> P-049: レシート明細データフレーム（df_receipt）の売上エポック秒（sales_epoch）を日付型（timestamp型）に変換し、"年"だけ取り出してレシート番号(receipt_no)、レシートサブ番号（receipt_sub_no）とともに抽出せよ。データは10件を抽出すれば良い。

In [ ]:
df49=df_receipt.copy()
df49['sales_epoch']=pd.to_datetime(df49['sales_epoch'])
df49['sales_epoch']=df49['sales_epoch'].dt.strftime('%Y')
df49[['receipt_no','receipt_sub_no','sales_epoch']].head(10)

#df49['sales_epoch']

receipt_no  receipt_sub_no sales_epoch
0         112               1        1970
1        1132               2        1970
2        1102               1        1970
3        1132               1        1970
4        1102               2        1970
5        1112               1        1970
6        1102               2        1970
7        1102               1        1970
8        1112               2        1970
9        1102               1        1970

---
> P-050: レシート明細データフレーム（df_receipt）の売上エポック秒（sales_epoch）を日付型（timestamp型）に変換し、"月"だけ取り出してレシート番号(receipt_no)、レシートサブ番号（receipt_sub_no）とともに抽出せよ。なお、"月"は0埋め2桁で取り出すこと。データは10件を抽出すれば良い。

In [ ]:
df50=df_receipt.copy()
df50['sales_epoch']=pd.to_datetime(df50['sales_epoch'])
df50['sales_epoch']=df50['sales_epoch'].dt.strftime('%m')
df50[['receipt_no','receipt_sub_no','sales_epoch']].head(10)

#df50['sales_epoch']

receipt_no  receipt_sub_no sales_epoch
0         112               1          01
1        1132               2          01
2        1102               1          01
3        1132               1          01
4        1102               2          01
5        1112               1          01
6        1102               2          01
7        1102               1          01
8        1112               2          01
9        1102               1          01

---
> P-051: レシート明細データフレーム（df_receipt）の売上エポック秒（sales_epoch）を日付型（timestamp型）に変換し、"日"だけ取り出してレシート番号(receipt_no)、レシートサブ番号（receipt_sub_no）とともに抽出せよ。なお、"日"は0埋め2桁で取り出すこと。データは10件を抽出すれば良い。

---
> P-052: レシート明細データフレーム（df_receipt）の売上金額（amount）を顧客ID（customer_id）ごとに合計の上、売上金額合計に対して2000円以下を0、2000円超を1に2値化し、顧客ID、売上金額合計とともに10件表示せよ。ただし、顧客IDが"Z"から始まるのものは非会員を表すため、除外して計算すること。

---
> P-053: 顧客データフレーム（df_customer）の郵便番号（postal_cd）に対し、東京（先頭3桁が100〜209のもの）を1、それ以外のものを0に２値化せよ。さらにレシート明細データフレーム（df_receipt）と結合し、全期間において買い物実績のある顧客数を、作成した2値ごとにカウントせよ。

---
> P-054: 顧客データデータフレーム（df_customer）の住所（address）は、埼玉県、千葉県、東京都、神奈川県のいずれかとなっている。都道府県毎にコード値を作成し、顧客ID、住所とともに抽出せよ。値は埼玉県を11、千葉県を12、東京都を13、神奈川県を14とすること。結果は10件表示させれば良い。

---
> P-055: レシート明細データフレーム（df_receipt）の売上金額（amount）を顧客ID（customer_id）ごとに合計し、その合計金額の四分位点を求めよ。その上で、顧客ごとの売上金額合計に対して以下の基準でカテゴリ値を作成し、顧客ID、売上金額と合計ともに表示せよ。カテゴリ値は上から順に1〜4とする。結果は10件表示させれば良い。
>
> - 最小値以上第一四分位未満
> - 第一四分位以上第二四分位未満
> - 第二四分位以上第三四分位未満
> - 第三四分位以上

---
> P-056: 顧客データフレーム（df_customer）の年齢（age）をもとに10歳刻みで年代を算出し、顧客ID（customer_id）、生年月日（birth_day）とともに抽出せよ。ただし、60歳以上は全て60歳代とすること。年代を表すカテゴリ名は任意とする。先頭10件を表示させればよい。

---
> P-057: 前問題の抽出結果と性別（gender）を組み合わせ、新たに性別×年代の組み合わせを表すカテゴリデータを作成せよ。組み合わせを表すカテゴリの値は任意とする。先頭10件を表示させればよい。

---
> P-058: 顧客データフレーム（df_customer）の性別コード（gender_cd）をダミー変数化し、顧客ID（customer_id）とともに抽出せよ。結果は10件表示させれば良い。

---
> P-059: レシート明細データフレーム（df_receipt）の売上金額（amount）を顧客ID（customer_id）ごとに合計し、合計した売上金額を平均0、標準偏差1に標準化して顧客ID、売上金額合計とともに表示せよ。標準化に使用する標準偏差は、不偏標準偏差と標本標準偏差のどちらでも良いものとする。ただし、顧客IDが"Z"から始まるのものは非会員を表すため、除外して計算すること。結果は10件表示させれば良い。

---
> P-060: レシート明細データフレーム（df_receipt）の売上金額（amount）を顧客ID（customer_id）ごとに合計し、合計した売上金額を最小値0、最大値1に正規化して顧客ID、売上金額合計とともに表示せよ。ただし、顧客IDが"Z"から始まるのものは非会員を表すため、除外して計算すること。結果は10件表示させれば良い。

---
> P-061: レシート明細データフレーム（df_receipt）の売上金額（amount）を顧客ID（customer_id）ごとに合計し、合計した売上金額を常用対数化（底=10）して顧客ID、売上金額合計とともに表示せよ。ただし、顧客IDが"Z"から始まるのものは非会員を表すため、除外して計算すること。結果は10件表示させれば良い。

---
> P-062: レシート明細データフレーム（df_receipt）の売上金額（amount）を顧客ID（customer_id）ごとに合計し、合計した売上金額を自然対数化(底=e）して顧客ID、売上金額合計とともに表示せよ。ただし、顧客IDが"Z"から始まるのものは非会員を表すため、除外して計算すること。結果は10件表示させれば良い。

---
> P-063: 商品データフレーム（df_product）の単価（unit_price）と原価（unit_cost）から、各商品の利益額を算出せよ。結果は10件表示させれば良い。

---
> P-064: 商品データフレーム（df_product）の単価（unit_price）と原価（unit_cost）から、各商品の利益率の全体平均を算出せよ。
ただし、単価と原価にはNULLが存在することに注意せよ。

---
> P-065: 商品データフレーム（df_product）の各商品について、利益率が30%となる新たな単価を求めよ。ただし、1円未満は切り捨てること。そして結果を10件表示させ、利益率がおよそ30％付近であることを確認せよ。ただし、単価（unit_price）と原価（unit_cost）にはNULLが存在することに注意せよ。

---
> P-066: 商品データフレーム（df_product）の各商品について、利益率が30%となる新たな単価を求めよ。今回は、1円未満を四捨五入すること（0.5については偶数方向の丸めで良い）。そして結果を10件表示させ、利益率がおよそ30％付近であることを確認せよ。ただし、単価（unit_price）と原価（unit_cost）にはNULLが存在することに注意せよ。

---
> P-067: 商品データフレーム（df_product）の各商品について、利益率が30%となる新たな単価を求めよ。今回は、1円未満を切り上げること。そして結果を10件表示させ、利益率がおよそ30％付近であることを確認せよ。ただし、単価（unit_price）と原価（unit_cost）にはNULLが存在することに注意せよ。

---
> P-068: 商品データフレーム（df_product）の各商品について、消費税率10%の税込み金額を求めよ。 1円未満の端数は切り捨てとし、結果は10件表示すれば良い。ただし、単価（unit_price）にはNULLが存在することに注意せよ。

---
> P-069: レシート明細データフレーム（df_receipt）と商品データフレーム（df_product）を結合し、顧客毎に全商品の売上金額合計と、カテゴリ大区分（category_major_cd）が"07"（瓶詰缶詰）の売上金額合計を計算の上、両者の比率を求めよ。抽出対象はカテゴリ大区分"07"（瓶詰缶詰）の購入実績がある顧客のみとし、結果は10件表示させればよい。

---
> P-070: レシート明細データフレーム（df_receipt）の売上日（sales_ymd）に対し、顧客データフレーム（df_customer）の会員申込日（application_date）からの経過日数を計算し、顧客ID（customer_id）、売上日、会員申込日とともに表示せよ。結果は10件表示させれば良い（なお、sales_ymdは数値、application_dateは文字列でデータを保持している点に注意）。

---
> P-071: レシート明細データフレーム（df_receipt）の売上日（sales_ymd）に対し、顧客データフレーム（df_customer）の会員申込日（application_date）からの経過月数を計算し、顧客ID（customer_id）、売上日、会員申込日とともに表示せよ。結果は10件表示させれば良い（なお、sales_ymdは数値、application_dateは文字列でデータを保持している点に注意）。1ヶ月未満は切り捨てること。

---
> P-072: レシート明細データフレーム（df_receipt）の売上日（sales_ymd）に対し、顧客データフレーム（df_customer）の会員申込日（application_date）からの経過年数を計算し、顧客ID（customer_id）、売上日、会員申込日とともに表示せよ。結果は10件表示させれば良い。（なお、sales_ymdは数値、application_dateは文字列でデータを保持している点に注意）。1年未満は切り捨てること。

---
> P-073: レシート明細データフレーム（df_receipt）の売上日（sales_ymd）に対し、顧客データフレーム（df_customer）の会員申込日（application_date）からのエポック秒による経過時間を計算し、顧客ID（customer_id）、売上日、会員申込日とともに表示せよ。結果は10件表示させれば良い（なお、sales_ymdは数値、application_dateは文字列でデータを保持している点に注意）。なお、時間情報は保有していないため各日付は0時0分0秒を表すものとする。

---
> P-074: レシート明細データフレーム（df_receipt）の売上日（sales_ymd）に対し、当該週の月曜日からの経過日数を計算し、売上日、当該週の月曜日付とともに表示せよ。結果は10件表示させれば良い（なお、sales_ymdは数値でデータを保持している点に注意）。

---
> P-075: 顧客データフレーム（df_customer）からランダムに1%のデータを抽出し、先頭から10件データを抽出せよ。

---
> P-076: 顧客データフレーム（df_customer）から性別（gender_cd）の割合に基づきランダムに10%のデータを層化抽出データし、性別ごとに件数を集計せよ。

---
> P-077: レシート明細データフレーム（df_receipt）の売上金額（amount）を顧客単位に合計し、合計した売上金額の外れ値を抽出せよ。ただし、顧客IDが"Z"から始まるのものは非会員を表すため、除外して計算すること。なお、ここでは外れ値を平均から3σ以上離れたものとする。結果は10件表示させれば良い。

---
> P-078: レシート明細データフレーム（df_receipt）の売上金額（amount）を顧客単位に合計し、合計した売上金額の外れ値を抽出せよ。ただし、顧客IDが"Z"から始まるのものは非会員を表すため、除外して計算すること。なお、ここでは外れ値を第一四分位と第三四分位の差であるIQRを用いて、「第一四分位数-1.5×IQR」よりも下回るもの、または「第三四分位数+1.5×IQR」を超えるものとする。結果は10件表示させれば良い。

---
> P-079: 商品データフレーム（df_product）の各項目に対し、欠損数を確認せよ。

---
> P-080: 商品データフレーム（df_product）のいずれかの項目に欠損が発生しているレコードを全て削除した新たなdf_product_1を作成せよ。なお、削除前後の件数を表示させ、前設問で確認した件数だけ減少していることも確認すること。

---
> P-081: 単価（unit_price）と原価（unit_cost）の欠損値について、それぞれの平均値で補完した新たなdf_product_2を作成せよ。なお、平均値について1円未満は四捨五入とし、0.5については偶数寄せでかまわない。補完実施後、各項目について欠損が生じていないことも確認すること。

---
> P-082: 単価（unit_price）と原価（unit_cost）の欠損値について、それぞれの中央値で補完した新たなdf_product_3を作成せよ。なお、中央値について1円未満は四捨五入とし、0.5については偶数寄せでかまわない。補完実施後、各項目について欠損が生じていないことも確認すること。

---
> P-083: 単価（unit_price）と原価（unit_cost）の欠損値について、各商品の小区分（category_small_cd）ごとに算出した中央値で補完した新たなdf_product_4を作成せよ。なお、中央値について1円未満は四捨五入とし、0.5については偶数寄せでかまわない。補完実施後、各項目について欠損が生じていないことも確認すること。

---
> P-084: 顧客データフレーム（df_customer）の全顧客に対し、全期間の売上金額に占める2019年売上金額の割合を計算せよ。ただし、販売実績のない場合は0として扱うこと。そして計算した割合が0超のものを抽出せよ。 結果は10件表示させれば良い。また、作成したデータにNAやNANが存在しないことを確認せよ。

---
> P-085: 顧客データフレーム（df_customer）の全顧客に対し、郵便番号（postal_cd）を用いて経度緯度変換用データフレーム（df_geocode）を紐付け、新たなdf_customer_1を作成せよ。ただし、複数紐づく場合は経度（longitude）、緯度（latitude）それぞれ平均を算出すること。


---
> P-086: 前設問で作成した緯度経度つき顧客データフレーム（df_customer_1）に対し、申込み店舗コード（application_store_cd）をキーに店舗データフレーム（df_store）と結合せよ。そして申込み店舗の緯度（latitude）・経度情報（longitude)と顧客の緯度・経度を用いて距離（km）を求め、顧客ID（customer_id）、顧客住所（address）、店舗住所（address）とともに表示せよ。計算式は簡易式で良いものとするが、その他精度の高い方式を利用したライブラリを利用してもかまわない。結果は10件表示すれば良い。

$$
緯度（ラジアン）：\phi \\
経度（ラジアン）：\lambda \\
距離L = 6371 * arccos(sin \phi_1 * sin \phi_2
+ cos \phi_1 * cos \phi_2 * cos(\lambda_1 − \lambda_2))
$$

---
> P-087:  顧客データフレーム（df_customer）では、異なる店舗での申込みなどにより同一顧客が複数登録されている。名前（customer_name）と郵便番号（postal_cd）が同じ顧客は同一顧客とみなし、1顧客1レコードとなるように名寄せした名寄顧客データフレーム（df_customer_u）を作成せよ。ただし、同一顧客に対しては売上金額合計が最も高いものを残すものとし、売上金額合計が同一もしくは売上実績の無い顧客については顧客ID（customer_id）の番号が小さいものを残すこととする。

---
> P-088: 前設問で作成したデータを元に、顧客データフレームに統合名寄IDを付与したデータフレーム（df_customer_n）を作成せよ。ただし、統合名寄IDは以下の仕様で付与するものとする。
>
> - 重複していない顧客：顧客ID（customer_id）を設定
> - 重複している顧客：前設問で抽出したレコードの顧客IDを設定

---
> P-閑話: df_customer_1, df_customer_nは使わないので削除する。

---
> P-089: 売上実績のある顧客に対し、予測モデル構築のため学習用データとテスト用データに分割したい。それぞれ8:2の割合でランダムにデータを分割せよ。

---
> P-090: レシート明細データフレーム（df_receipt）は2017年1月1日〜2019年10月31日までのデータを有している。売上金額（amount）を月次で集計し、学習用に12ヶ月、テスト用に6ヶ月のモデル構築用データを3セット作成せよ。

---
> P-091: 顧客データフレーム（df_customer）の各顧客に対し、売上実績のある顧客数と売上実績のない顧客数が1:1となるようにアンダーサンプリングで抽出せよ。

---
> P-092: 顧客データフレーム（df_customer）では、性別に関する情報が非正規化の状態で保持されている。これを第三正規化せよ。

---
> P-093: 商品データフレーム（df_product）では各カテゴリのコード値だけを保有し、カテゴリ名は保有していない。カテゴリデータフレーム（df_category）と組み合わせて非正規化し、カテゴリ名を保有した新たな商品データフレームを作成せよ。

---
> P-094: 先に作成したカテゴリ名付き商品データを以下の仕様でファイル出力せよ。なお、出力先のパスはdata配下とする。
>
> - ファイル形式はCSV（カンマ区切り）
> - ヘッダ有り
> - 文字コードはUTF-8

---
> P-095: 先に作成したカテゴリ名付き商品データを以下の仕様でファイル出力せよ。なお、出力先のパスはdata配下とする。
>
> - ファイル形式はCSV（カンマ区切り）
> - ヘッダ有り
> - 文字コードはCP932

---
> P-096: 先に作成したカテゴリ名付き商品データを以下の仕様でファイル出力せよ。なお、出力先のパスはdata配下とする。
>
> - ファイル形式はCSV（カンマ区切り）
> - ヘッダ無し
> - 文字コードはUTF-8

---
> P-097: 先に作成した以下形式のファイルを読み込み、データフレームを作成せよ。また、先頭10件を表示させ、正しくとりまれていることを確認せよ。
>
> - ファイル形式はCSV（カンマ区切り）
> - ヘッダ有り
> - 文字コードはUTF-8

---
> P-098: 先に作成した以下形式のファイルを読み込み、データフレームを作成せよ。また、先頭10件を表示させ、正しくとりまれていることを確認せよ。
>
> - ファイル形式はCSV（カンマ区切り）
> - ヘッダ無し
> - 文字コードはUTF-8

---
> P-099: 先に作成したカテゴリ名付き商品データを以下の仕様でファイル出力せよ。なお、出力先のパスはdata配下とする。
>
> - ファイル形式はTSV（タブ区切り）
> - ヘッダ有り
> - 文字コードはUTF-8

---
> P-100: 先に作成した以下形式のファイルを読み込み、データフレームを作成せよ。また、先頭10件を表示させ、正しくとりまれていることを確認せよ。
>
> - ファイル形式はTSV（タブ区切り）
> - ヘッダ有り
> - 文字コードはUTF-8

# これで１００本終わりです。おつかれさまでした！